### ISSUE TITLE
No error raised from calling an object annotated as a dict

### ISSUE THREAD

**Issue by ajhynes7** [python/mypy#12761]
**Bug Report**

In the function below, mypy raises an error because a dict is not callable.

```python
def func(arg: dict):
    arg()
```

If the function includes a `callable` check, mypy does not raise an error. But I expected that mypy would still raise an error because the argument is only annotated as a dict.

```python
def func(arg: dict):

    if callable(arg): 
        arg()
```

I'd expect mypy to run successfully only if the argument was annotated as a union of a dict and a callable, like this:

```python
from typing import Callable


def func(arg: dict | Callable):

    if callable(arg):
        arg()

```

**Your Environment**

- Mypy version used: 0.950
- Python version used: 3.10.0

**Comment by AlexWaygood** [comment]
Mypy's behaviour is correct here. A function argument annotated with `dict` can be passed any subtype of `dict`, and there's an infinite variety of plausible subtypes of `dict` that would be callable. E.g. this works fine at runtime:

```python
class CallableDict(dict):
    def __call__(self) -> str:
        return 'Called'

def func(arg: dict):
    if callable(arg): 
        arg()

func(CallableDict({1: 2}))
```
